### ISSUE TITLE
Support providing endpoint from a container JAR

### ISSUE THREAD

**Issue by glassfishrobot** [jakartaee/websocket#240]
For JSF 2.3 / Mojarra I'm currently in progress of adding a standard <f:websocket> tag. As Mojarra is to be designed as a container-provided JAR (not an user-provided JAR), the endpoint needs to be programmatically added in a ServletContextListener, but at that point, servletContext.getAttribute(ServerContainer.class.getName()) returns null when Tyrus is used (Payara/GlassFish). Upon inspection it turns out that the TyrusServletContainerInitializer#onStartup() immediately returns when there are no user-provided endpoints and doesn't register the ServerContainer anymore, causing it to not be placed in ServletContext anymore. See also [https://java.net/jira/browse/TYRUS-420](https://java.net/jira/browse/TYRUS-420)

I'm therefore making the following proposals to ensure that the websocket container can forcibly be initialized from the container on regardless of the user-provided endpoints.

1\. Proposal #1: require eager initialization.

*   Always initialize container regardless of the onStartup(classes).

2\. Proposal #2: check a context param.

*   Check if a context param something like "javax.websocket.ALWAYS_ENABLED" is set.
*   If so, initialize container regardless of the onStartup(classes).

3\. Proposal #3: check a context attribute.

*   Check if servletContext.getAttribute(ServerContainer.class.getName()) returns null.
*   If so, check if ServerContainer.class.getName() is present in servletContext.getAttributeNames().
*   If so, initialize container regardless of the onStartup(classes).

**Comment by markt-asf** [comment]
Currently section 6.2 of the spec says:

> Java EE containers are not required to support deployment of websocket endpoints if they are not packaged in a WAR file as described above.

I think the simplest option is a version of option 1. Essentially, if the SCI is executed then the `ServerContainer` is always added to the `ServletContext`. The Servlet specification provides a mechanism for an application developer to disable an SCI the developer does not want to be load (8.2.2. b).

**Comment by joakime** [comment]
This is the first time i've seen this particular issue.

Just thinking, wouldn't a container implementation of `javax.websocket.server.ServerApplicationConfig` be sufficient to have Mojarra auto-register it's endpoints?

**Comment by markt-asf** [comment]
It depends on what the WebSocket implementation scans. Section 6.2 of the WebSocket spec allows the implementation to limit itself just to the current WAR. As does the Servlet spec when it comes to annotation scanning. However the Servlet spec does indicate that SCIs can be provided by the container. And, contrary to what I wrote above, there is no spec provided mechanism (there may be be a container specific one) to exclude container provided SCIs.

I'm tempted to still go with the variation I proposed above. If the SCI executes, the `ServerContainer` is added to the `ServletContext`. If that is a problem, containers can provide container specific mechanisms to exclude the SCI for some/all web applications. It seems like the simplest option.
